### ISSUE TITLE
Negative value on measurement result

### ISSUE THREAD

**Issue by sanoylab** [m-lab/measure-saver#20]
We are using the Measurement Saver to test the internet connectivity of schools and have noticed that sometimes the plugin returns negative values for latency and download speed. Can please fix this issue or explain why negative values are being returned?

**Comment by robertodauria** [comment]
Hi @sanoylab, have you seen this in the browser plugin's UI, in the database, or both? Could you share more details? A complete row (perhaps anonymized?) would help.

**Comment by sanoylab** [comment]
Thanks, @robertodauria for the quick reply. We saw the issue on the database. So basically we have a desktop app installed on schools that runs multiple time in the background and send us the internet connectivity speed to our database. And now we noticed a negative value for the speed download metics. Please see the sample data:
`[
  {
    "id": 0000,
    "Timestamp": "2022-12-23 04:03:13.099+00",
    "UUID": "ndt-85gdq_1671049653_0000000000117581",
    "BrowserID": "eefa0f3b-9d97-43f9-b5c5-23ec2b87397b",
    "school_id": XXXXXXXXXXX,
    "DeviceType": "windows",
    "Notes": "",
    "ClientInfo": {
      "IP": "000000000",
      "ASN": "AS819",
      "ISP": "XXXXXXXXXXXX",
      "City": "XXXXXXXXXXXX",
      "Region": "XXXXXXXXXXXX",
      "Country": "XXXXXXXXXXXX",
      "Latitude": 0000,
      "Timezone": "Asia/Tashkent",
      "Longitude": 0000
    },
    "ServerInfo": {
      "URL": "http://ndt-iupui-mlab2-del02.mlab-oti.measurement-lab.org:7123",
      "City": "New Delhi",
      "FQDN": "ndt-iupui-mlab2-del02.mlab-oti.measurement-lab.org",
      "IPv4": "00000000",
      "Site": "del02",
      "Label": "New Delhi",
      "Metro": "del",
      "Country": "IN"
    },
    "annotation": "NULL",
    "Download": -4.94850849599456,
    "Upload": 4899,
    "Latency": 243,
    "Results": {
      "CurMSS": "1424",
      "CurRTO": "0",
      "MaxRTT": "338",
      "MinRTT": "243",
      "Sndbuf": "0",
      "SumRTT": "26065",
      "MaxCwnd": "0",
      "PktsOut": "30164",
      "c2sRate": 4899,
      "s2cRate": -4.948508495994557,
      "CountRTT": "102",
      "Timeouts": "0",
      "AckPktsIn": "0",
      "DupAcksIn": "0",
      "MaxRwinRcvd": "0",
      "PktsRetrans": "0",
      "RcvWinScale": "0",
      "SndWinScale": "0",
      "TCPInfo.ATO": "40000",
      "TCPInfo.RTO": "540000",
      "TCPInfo.RTT": "338200",
      "DataBytesOut": "0",
      "TCPInfo.Lost": "0",
      "TCPInfo.PMTU": "1500",
      "TCPInfo.State": "1",
      "SndLimTimeCwnd": "0",
      "SndLimTimeRwin": "0",
      "TCPInfo.AdvMSS": "1448",
      "TCPInfo.MinRTT": "242889",
      "TCPInfo.Probes": "0",
      "TCPInfo.RTTVar": "621",
      "TCPInfo.RcvMSS": "596",
      "TCPInfo.RcvRTT": "245000",
      "TCPInfo.Sacked": "0",
      "TCPInfo.SegsIn": "13624",
      "TCPInfo.SndMSS": "1424",
      "TCPInfo.SndWnd": "9506560",
      "TCPInfo.WScale": "120",
      "TCPInfo.Backoff": "0",
      "TCPInfo.CAState": "0",
      "TCPInfo.Fackets": "0",
      "TCPInfo.Options": "7",
      "TCPInfo.Retrans": "0",
      "TCPInfo.SegsOut": "30164",
      "TCPInfo.SndCwnd": "1752",
      "TCPInfo.Unacked": "1748",
      "SndLimTimeSender": "0",
      "TCPInfo.BusyTime": "10104000",
      "TCPInfo.RcvSpace": "14600",
      "CongestionSignals": "0",
      "TCPInfo.BytesSent": "42949242",
      "TCPInfo.DSackDups": "0",
      "TCPInfo.Delivered": "28416",
      "TCPInfo.ReordSeen": "3396",
      "NDTResult.S2C.UUID": "ndt-85gdq_1671049653_0000000000117581",
      "TCPInfo.AppLimited": "0",
      "TCPInfo.BytesAcked": "40461694",
      "TCPInfo.DataSegsIn": "1",
      "TCPInfo.PacingRate": "8852187",
      "TCPInfo.RcvOooPack": "0",
      "TCPInfo.Reordering": "5",
      "NDTResult.S2C.Error": "",
      "TCPInfo.DataSegsOut": "30163",
      "TCPInfo.DeliveredCE": "0",
      "TCPInfo.LastAckRecv": "4",
      "TCPInfo.LastAckSent": "0",
      "TCPInfo.RWndLimited": "320000",
      "TCPInfo.RcvSsThresh": "64076",
      "TCPInfo.Retransmits": "0",
      "TCPInfo.SndSsThresh": "368",
      "NDTResult.S2C.MaxRTT": "338ms",
      "NDTResult.S2C.MinRTT": "243ms",
      "NDTResult.S2C.SumRTT": "26.065s",
      "TCPInfo.BytesRetrans": "0",
      "TCPInfo.DeliveryRate": "7004874",
      "TCPInfo.LastDataRecv": "10104",
      "TCPInfo.LastDataSent": "4",
      "TCPInfo.NotsentBytes": "175152",
      "TCPInfo.TotalRetrans": "0",
      "NDTResult.S2C.EndTime": "2022-12-22 08:41:37.669918825 +0000 UTC m=+648556.122715694",
      "TCPInfo.BytesReceived": "596",
      "TCPInfo.MaxPacingRate": "-1",
      "TCPInfo.SndBufLimited": "0",
      "packetRetransmissions": 0,
      "NDTResult.S2C.ClientIP": "000000000",
      "NDTResult.S2C.CountRTT": "102",
      "NDTResult.S2C.ServerIP": "00000000",
      "NDTResult.S2C.StartTime": "2022-12-22 08:41:27.533143996 +0000 UTC m=+648545.985940885",
      "NDTResult.S2C.ClientPort": "0000",
      "NDTResult.S2C.ServerPort": "0000"
    },
    
  }
]`
As you can see above the 'Download' value is a negative value.

**Comment by robertodauria** [comment]
How many instances of this have you found?

The most likely explanation, I believe, is that the system's clock was updated during the measurement (likely an NTP daemon, i.e. Windows' date/time is set to "automatic") so that the new "current time" changed to before the measurement's start time, which causes a negative speed (`throughput = bytesReceived / (currentTime - startTime)` so if `currentTime < startTime` the throughput will appear to be negative).

This seems particularly likely if the machine was just back from standby/hibernation and Chrome's background task scheduling restarted more or less at the same time Windows updated its local time with NTP.

There's a possible fix for the browser extension that I'm exploring. Also, the backend should probably validate that the Download and Upload fields aren't negative, but I was also curious to know how widespread the problem is.

**Comment by sanoylab** [comment]
I think this is the second time happened. A while back we also got a negative value from the browser extension. 

Your logic makes sense if `throughput = bytesReceived / (currentTime - startTime` and clock change might create a negative value.

Thank you for taking the time to take a look at a solution. On our side for now we added a validation rule before saving the data (measurement speed values > 0)
